### PR TITLE
feat: builder plugin image compress

### PR DESCRIPTION
**Pull request by Asuka109**
# PR Details

## Description

**New external plugin**: `@modern-js/webpack-builder-plugin-image-compress`

Refer to https://github.com/modern-js-dev/modern.js/pull/1548

## TODOs

- [ ] `@modern-js/webpack-builder` exports e2e testing utils

## Types of changes

- [ ] Docs change / refactoring / dependency upgrade
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [ ] My code follows the code style of this project.
- [ ] I have updated changeset
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: packages/builder/plugin-image-compress/package.json [hunk 1]
```diff
@@ -0,0 +1,65 @@
+{
+  "private": true,
```
Comment (chenjiahan) [inline packages/builder/plugin-image-compress/package.json:2]: private 可以去掉咯

File: packages/builder/plugin-image-compress/package.json [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+  "keywords": [
+    "react",
+    "framework",
+    "modern",
+    "modern.js"
+  ],
+  "engines": {
+    "node": ">=14.0.0"
+  },
+  "version": "0.0.1",
```
Comment (chenjiahan) [inline packages/builder/plugin-image-compress/package.json:18]: 版本对齐 next 分支上的最新版号

File: packages/builder/plugin-image-compress/package.json [hunk 3]
```diff
@@ -0,0 +1,65 @@
+{
+  "private": true,
+  "name": "@modern-js/builder-plugin-image-compress",
```
Comment (chenjiahan) [inline packages/builder/plugin-image-compress/package.json:3]: 这个插件的文档也补充一下哈

File: packages/builder/plugin-image-compress/src/index.ts [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+/** Options enable by default: {@link DEFAULT_OPTIONS} */
+export const PluginImageCompress = (...options: Options[]): BuilderPlugin => ({
+  name: 'builder-plugin-image-compress',
+  setup(api) {
+    const optsWithDefault = options.length ? options : DEFAULT_OPTIONS;
+    const opts = optsWithDefault.map(opt => withDefaultOptions(opt));
+    api.modifyWebpackChain((chain, { CHAIN_ID }) => {
+      _.each(opts, opt => {
+        chain.module
+          .rule(CHAIN_ID.RULE.IMAGE)
```
Comment (chenjiahan) [inline packages/builder/plugin-image-compress/src/index.ts:19]: CHAIN_ID.RULE.IMAGE -> CHAIN_ID.RULE.IMAGE_COMPRESS

File: packages/builder/plugin-image-compress/src/index.ts [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+          .use(`web-builder-plugin-image-compress#${opt.use}`)
```
Comment (chenjiahan) [inline packages/builder/plugin-image-compress/src/index.ts:20]: loader 名字里没必要带上 builder-plugin

`web-builder-plugin-image-compress` -> `image-compress`

File: packages/builder/plugin-image-compress/src/index.ts [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+export const loaderPath = path.resolve(__dirname, 'loader.js');
+export const DEFAULT_OPTIONS: Codecs[] = ['jpeg', 'png', 'ico'];
+
+/** Options enable by default: {@link DEFAULT_OPTIONS} */
+export const PluginImageCompress = (...options: Options[]): BuilderPlugin => ({
+  name: 'builder-plugin-image-compress',
+  setup(api) {
+    const optsWithDefault = options.length ? options : DEFAULT_OPTIONS;
+    const opts = optsWithDefault.map(opt => withDefaultOptions(opt));
+    api.modifyWebpackChain((chain, { CHAIN_ID }) => {
```
Comment (chenjiahan) [inline packages/builder/plugin-image-compress/src/index.ts:16]: dev 的时候是不是不应该做压缩，影响编译性能

File: packages/builder/plugin-image-compress/src/index.ts (hunk 4 above)
Comment (Asuka109) [reply packages/builder/plugin-image-compress/src/index.ts:19]: 我一开始想的是 compressor 就放在 image rule 里，现在看还是得拆成单独的 rule... 容我再改改

File: packages/builder/plugin-image-compress/src/index.ts (hunk 6 above)
Comment (Asuka109) [reply packages/builder/plugin-image-compress/src/index.ts:16]: 目前的配置格式是：

```ts
Plugin([
  { use: 'jpeg' },
  { use: 'png' }
])
```

这种形式可能没法给出一个统一的开关去控制是否启用，具体地说用户没有办法通过类似 enable 的属性强制在 dev 模式启用压缩。如果没有这种需求的话我就内部写死用环境变量当作开关如何🤔

File: packages/builder/plugin-image-compress/package.json (hunk 3 above)
Comment (Asuka109) [reply packages/builder/plugin-image-compress/package.json:3]: okk

File: packages/builder/plugin-image-compress/src/index.ts (hunk 4 above)
Comment (Asuka109) [reply packages/builder/plugin-image-compress/src/index.ts:19]: 这样做的话好像得添加好几条 rule name，我放插件内部维护吧

```ts

export const CHAIN_ID = {
  RULE: {
    IMAGE_COMPRESS_JPEG: 'image-compress-jpeg',
    IMAGE_COMPRESS_PNG: 'image-compress-png',
  },
} as const;

/** Options enable by default: {@link DEFAULT_OPTIONS} */
export const PluginImageCompress = (...options: Options[]): BuilderPlugin => ({
  name: 'builder-plugin-image-compress',
  setup(api) {
    const optsWithDefault = options.length ? options : DEFAULT_OPTIONS;
    const opts = optsWithDefault.map(opt => withDefaultOptions(opt));
    api.modifyWebpackChain((chain) => {
      _.each(opts, opt => {
        chain.module
          .rule(CHAIN_ID.RULE.IMAGE_COMPRESS_XXX)
          .use(`image-compress#${opt.use}`)
          .loader(path.resolve(__dirname, './loader'))
          .options(opt);
      });
    });
  },
});
```

File: packages/builder/plugin-image-compress/package.json (hunk 3 above)
Comment (Asuka109) [reply packages/builder/plugin-image-compress/package.json:2]: 补充好了 有空看下哈~

**Review by chenjiahan — APPROVED** [review]

**Pull request merged**